custom_metrics/responsive_images.js

```diff
@@ -0,0 +1,895 @@
+[responsive-images]
+// Uncomment the previous line for testing on webpagetest.org
+
+return JSON.stringify({
+	'responsive-images': (() => {
+
+		const imgs = [ ...document.querySelectorAll( 'img' ) ];
+		return imgs.map( ( img ) => getImgData( img ) );
+
+	})(),
+});
+
+
+
+// take a CSS length (as a String)
+// return a Number of CSS pixels
+//
+function resolveLength( cssLengthString ) {
+
+	let cleanup = false,
+	    myHTML;
+
+	if ( !document.documentElement ) {
+		myHTML = document.createElement( 'html' );
+		document.appendChild( myHTML );
+		cleanup = true;
+	}
+
+	const testDiv = document.createElement( 'div' );
+	testDiv.style.fontSize = 'initial';
+	testDiv.style.boxSizing = 'border-box';
+	document.documentElement.appendChild( testDiv );
+	testDiv.style.width = cssLengthString;
+	const width = testDiv.getBoundingClientRect().width;
+
+	testDiv.remove();
+	if ( cleanup ) {
+		myHTML.remove();
+	}
+
+	return width;
+}
+
+// determine if the user agent will pick a <source> with a given type="${mimeTypeString}"
+// takes a String
+//
+// I was trying to do this the "right" way, but async code and promises got involved, and I give up and am going to hardcode
+// source: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/common/mime_util/mime_util.cc;l=164;drc=c14f6f4b9c44fe479a8d004576b42723b2a5feb6;bpv=0;bpt=1
+
+function matchType( mimeTypeString ) {
+
+	const supportedTypes = [
+		"image/jpeg",
+		"image/pjpeg",
+		"image/jpg",
+		"image/webp",
+		"image/png",
+		"image/apng",
+		"image/gif",
+		"image/bmp",
+		"image/vnd.microsoft.icon",  // ico
+		"image/x-icon",              // ico
+		"image/x-xbitmap",           // xbm
+		"image/x-png",
+		"image/avif" // enabled for both desktop and mobile now (2021-06-29), phew
+	];
+	
+	return supportedTypes.includes( mimeTypeString.toLowerCase() );
+
+}
+
+// the candidate <source> elements of a given <picture>
+// takes a <picture> Element
+// returns a plain ol' Array of <source> Elements
+function pictureSources( picture ) {
+	// all <source>s that come BEFORE the <img>
+	// https://twitter.com/zcorpan/status/1409778636153032706
+	return [ ...picture.querySelectorAll( 'source:not( img ~ source )' ) ];
```

Comment:
Maybe this is overly pedantic, but this can still return incorrect `source` elements (e.g. grandchildren).

I would pass in the `img` as argument, and walk to the `parentNode` then through its `children` until you find the `img` that was passed in.

Reply:
Done!